### PR TITLE
Ros2 v0.8.0 autoware debug tools

### PR DESCRIPTION
**Pull request by sgk-000**

### CODE REVIEW

**Comment by tkimura4** [comment]
Could you place python scripts in the "scripts" folder instead of the "src" folder？

**Comment by tkimura4** [comment]
common/util/autoware_debug_tools/__init__.py　->　Is this unnecessary file?

**Comment by tkimura4** [comment]
Could you add user execute permission to python files?
(sudo chmod u+x ~~~.py)

File: common/util/autoware_debug_tools/src/pose2tf.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        super().__init__('pose2tf')
+        self._options = options
+        self._tf_broadcaster = tf2_ros.TransformBroadcaster(self)
+        self._sub_pose = self.create_subscription(
+            PoseStamped, self._options.topic_name, self._on_pose, 1)
+
+    def _on_pose(self, msg):
+        try:
+            (trans, quat) = Pose2TfNode.create_transform(msg.pose)
+            self._tf_broadcaster.sendTransform(
```
Comment (tkimura4) [inline common/util/autoware_debug_tools/src/pose2tf.py:38]: sendTransform function has only 1 argument.
Plaese check the following code.
/opt/ros/foxy/lib/python3.8/site-packages/tf2_ros/transform_broadcaster.py

This is C++ code, but it's also helpful
https://github.com/tier4/AutowareArchitectureProposal.iv/blob/master/localization/pose_twist_fusion_filter/ekf_localizer/src/ekf_localizer.cpp#L200-L213

File: common/util/autoware_debug_tools/src/self_pose_listener.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+from tf2_ros.buffer import Buffer
+from tf2_ros.transform_listener import TransformListener
+from geometry_msgs.msg import PoseStamped
+
+
+class SelfPoseListener(Node):
+    def __init__(self):
+        super().__init__('self_pose_listener')
+        self.tf_buffer = Buffer()
+        self.tf_listener = TransformListener(self.tf_buffer, self)
```
Comment (tkimura4) [inline common/util/autoware_debug_tools/src/self_pose_listener.py:29]: self.tf_listener -> self._tf_listener

File: common/util/autoware_debug_tools/src/self_pose_listener.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+    def get_current_pose(self):
+        try:
+            (trans, quat) = self.tf_buffer.lookup_transform("map", "base_link", rclpy.time.Time())
+            time = self._tf_listener.getLatestCommonTime("map", "base_link")
```
Comment (tkimura4) [inline common/util/autoware_debug_tools/src/self_pose_listener.py:34]: TransformListener does not have getLatestCommonTime method.

File: common/util/autoware_debug_tools/src/tf2pose.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        self.tf_listener = TransformListener(self.tf_buffer, self)
+
+        self._pub_pose = self.create_publisher(PoseStamped, "pose", 1)
+        self.timer = self.create_timer((1.0 / self._options.hz), self._on_timer)
+
+    def _on_timer(self):
+        try:
+            (trans, quat) = self.tf_buffer.lookup_transform(
+                self._options.tf_from, self._options.tf_to, rclpy.time.Time())
+            time = self._tf_listener.getLatestCommonTime(
```
Comment (tkimura4) [inline common/util/autoware_debug_tools/src/tf2pose.py:46]: TransformListener does not have getLatestCommonTime method.

File: common/util/autoware_debug_tools/src/stop_reason2pose.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+                    snake_case_stop_reason, pose.pose, th_dist)
+                if nearest_pose_id:
+                    self._update_pose(snake_case_stop_reason, pose.pose, nearest_pose_id)
+                    pose_id = nearest_pose_id
+                else:
+                    pose_id = self._register_pose(snake_case_stop_reason, pose.pose)
+
+                pose_topic_name = "{snake_case_stop_reason}_{pose_id}".format(**locals())
+                if pose_topic_name not in self._pub_pose_map:
+                    self._pub_pose_map[pose_topic_name] = self.create_publisher(
```
Comment (tkimura4) [inline common/util/autoware_debug_tools/src/stop_reason2pose.py:66]: create_publisher has not "queue_size" arugument

File: common/util/autoware_debug_tools/src/stop_reason2pose.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+            # Publish nearest stop_reason without number
+            self_pose = self._self_pose_listener.get_current_pose()
+
+            nearest_pose = PoseStamped()
+            nearest_pose.header = msg.header
+            nearest_pose.pose = self._get_nearest_pose_in_array(stop_reason, self_pose.pose)
+
+            if nearest_pose.pose:
+                if snake_case_stop_reason not in self._pub_pose_map:
+                    self._pub_pose_map[snake_case_stop_reason] = self.create_publisher(
```
Comment (tkimura4) [inline common/util/autoware_debug_tools/src/stop_reason2pose.py:79]: create_publisher has not "queue_size" arugument

**Comment by sgk-000** [comment]
test tf2pose on psim
`ros2 run autoware_debug_tools tf2pose.py base_link camera0/camera_link 100`
`ros2 topic echo /pose`
```
---
header:
  stamp:
    sec: 0
    nanosec: 0
  frame_id: base_link
pose:
  position:
    x: 1.023561000869848
    y: 0.5590264239645132
    z: 1.720888340849957
  orientation:
    x: -0.08256225058667584
    y: 0.13838973437946628
    z: 0.47517449624901736
    w: 0.8650092220958897
---

```

**Comment by sgk-000** [comment]
test pose2tf on psim
`ros2 run autoware_debug_tools tf2pose.py base_link camera0/camera_link 100`
`ros2 run autoware_debug_tools pose2tf.py /pose pose_link`
`ros2 run tf2_ros tf2_echo pose_link base_link`
```
At time 0.0
- Translation: [-0.422, 0.585, -1.950]
- Rotation: in Quaternion [0.083, -0.138, -0.475, 0.865]
```

**Comment by sgk-000** [comment]
test stop_reason2pose on psim
```
ros2 run autoware_debug_tools stop_reason2pose.py  /planning/scenario_planning/status/stop_reasons
ros2 topic echo /autoware_debug_tools/stop_reason2pose/obstacle_stop
header:
  stamp:
    sec: 1614178534
    nanosec: 665268145
  frame_id: map
pose:
  position:
    x: 3697.4268355476393
    y: 73762.90924363749
    z: -0.7630358717756539
  orientation:
    x: 0.0
    y: 0.0
    z: 0.2382767498066294
    w: 0.9711972974126261
---
```

**Comment by sgk-000** [comment]
test stop_reason2tf on psim
```
 ros2 run autoware_debug_tools stop_reason2tf obstacle_stop
At time 1614179492.158942809
- Translation: [1.373, 0.565, -0.025]
- Rotation: in Quaternion [0.000, -0.000, 0.007, 1.000]
```

File: common/util/autoware_debug_tools/src/pose2tf.py (hunk 1 above)
Comment (sgk-000) [reply common/util/autoware_debug_tools/src/pose2tf.py]: fixed!

File: common/util/autoware_debug_tools/src/self_pose_listener.py (hunk 2 above)
Comment (sgk-000) [reply common/util/autoware_debug_tools/src/self_pose_listener.py]: fixed!

File: common/util/autoware_debug_tools/src/self_pose_listener.py (hunk 3 above)
Comment (sgk-000) [reply common/util/autoware_debug_tools/src/self_pose_listener.py]: fixed!

File: common/util/autoware_debug_tools/src/tf2pose.py (hunk 4 above)
Comment (sgk-000) [reply common/util/autoware_debug_tools/src/tf2pose.py]: fixed!

File: common/util/autoware_debug_tools/src/stop_reason2pose.py (hunk 5 above)
Comment (sgk-000) [reply common/util/autoware_debug_tools/src/stop_reason2pose.py]: fixed!

File: common/util/autoware_debug_tools/src/stop_reason2pose.py (hunk 6 above)
Comment (sgk-000) [reply common/util/autoware_debug_tools/src/stop_reason2pose.py]: fixed!

**Comment by tkimura4** [comment]
ros2 run autoware_debug_tools tf2pose.py viewer map 100
ros2 run autoware_debug_tools pose2tf.py /autoware_debug_tools/tf2pose/pose pose_link
ros2 topic echo /autoware_debug_tools/tf2pose/pose
ros2 topic echo /tf
-> OK

**Comment by tkimura4** [comment]
LGTM

**Review by tkimura4 — APPROVED** [review]

**Pull request merged**